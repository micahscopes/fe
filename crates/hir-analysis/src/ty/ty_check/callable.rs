use hir::{
    hir_def::{CallArg as HirCallArg, ExprId, GenericArgListId, IdentId},
    span::{
        expr::{LazyCallArgListSpan, LazyCallArgSpan},
        params::LazyGenericArgListSpan,
        DynLazySpan,
    },
};
use if_chain::if_chain;

use super::{ExprProp, TyChecker};
use crate::{
    ty::{
        diagnostics::{BodyDiag, FuncBodyDiag},
        fold::{TyFoldable, TyFolder},
        func_def::FuncDef,
        ty_def::{TyBase, TyData, TyId},
        ty_lower::lower_generic_arg_list,
        visitor::{TyVisitable, TyVisitor},
    },
    HirAnalysisDb,
};

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Callable<'db> {
    func_def: FuncDef<'db>,
    generic_args: Vec<TyId<'db>>,
}

impl<'db> TyVisitable<'db> for Callable<'db> {
    fn visit_with<V>(&self, visitor: &mut V)
    where
        V: TyVisitor<'db>,
    {
        self.generic_args.visit_with(visitor)
    }
}

impl<'db> TyFoldable<'db> for Callable<'db> {
    fn super_fold_with<F>(self, folder: &mut F) -> Self
    where
        F: TyFolder<'db>,
    {
        Self {
            func_def: self.func_def,
            generic_args: self.generic_args.fold_with(folder),
        }
    }
}

impl<'db> Callable<'db> {
    pub(super) fn new(
        db: &'db dyn HirAnalysisDb,
        ty: TyId<'db>,
        span: DynLazySpan<'db>,
    ) -> Result<Self, FuncBodyDiag<'db>> {
        let (base, args) = ty.decompose_ty_app(db);

        if base.is_ty_var(db) {
            return Err(BodyDiag::TypeMustBeKnown(span).into());
        }

        let TyData::TyBase(TyBase::Func(func_def)) = base.data(db) else {
            return Err(BodyDiag::not_callable(db, span, ty).into());
        };

        let params = ty.generic_args(db);
        assert_eq!(params.len(), args.len());

        Ok(Self {
            func_def: *func_def,
            generic_args: args.to_vec(),
        })
    }

    pub fn ret_ty(&self, db: &'db dyn HirAnalysisDb) -> TyId<'db> {
        self.func_def.ret_ty(db).instantiate(db, &self.generic_args)
    }

    pub fn ty(&self, db: &'db dyn HirAnalysisDb) -> TyId<'db> {
        let ty = TyId::func(db, self.func_def);
        TyId::foldl(db, ty, &self.generic_args)
    }

    pub(super) fn unify_generic_args(
        &mut self,
        tc: &mut TyChecker<'db>,
        args: GenericArgListId<'db>,
        span: LazyGenericArgListSpan<'db>,
    ) -> bool {
        let db = tc.db;
        let hir_db = db.as_hir_db();

        if !args.is_given(hir_db) {
            return true;
        }

        let given_args = lower_generic_arg_list(db, args, tc.env.scope());
        let offset = self.func_def.offset_to_explicit_params_position(db);
        let current_args = &mut self.generic_args[offset..];

        if current_args.len() != given_args.len() {
            let diag = BodyDiag::CallGenericArgNumMismatch {
                primary: span.into(),
                def_span: self.func_def.name_span(db),
                given: given_args.len(),
                expected: current_args.len(),
            };
            tc.push_diag(diag);

            return false;
        }

        for (i, (&given, arg)) in given_args.iter().zip(current_args.iter_mut()).enumerate() {
            *arg = tc.equate_ty(given, *arg, span.arg(i).into());
        }

        true
    }

    pub(super) fn check_args(
        &self,
        tc: &mut TyChecker<'db>,
        call_args: &[HirCallArg<'db>],
        span: LazyCallArgListSpan<'db>,
        receiver: Option<(ExprId, ExprProp<'db>)>,
    ) {
        let db = tc.db;

        let expected_arity = self.func_def.arg_tys(db).len();
        let given_arity = if receiver.is_some() {
            call_args.len() + 1
        } else {
            call_args.len()
        };
        if given_arity != expected_arity {
            let diag = BodyDiag::CallArgNumMismatch {
                primary: span.into(),
                def_span: self.func_def.name_span(db),
                given: given_arity,
                expected: expected_arity,
            };
            tc.push_diag(diag);
            return;
        }

        let mut args = if let Some((receiver_expr, receiver_prop)) = receiver {
            let mut args = Vec::with_capacity(call_args.len() + 1);
            let arg = CallArg::new(
                IdentId::make_self(db.as_hir_db()).into(),
                receiver_prop,
                None,
                receiver_expr.lazy_span(tc.body()).into(),
            );
            args.push(arg);
            args
        } else {
            Vec::with_capacity(call_args.len())
        };

        for (i, hir_arg) in call_args.iter().enumerate() {
            let arg = CallArg::from_hir_arg(tc, hir_arg, span.arg(i));
            args.push(arg);
        }

        for (i, (given, expected)) in args
            .into_iter()
            .zip(self.func_def.arg_tys(db).iter())
            .enumerate()
        {
            if_chain! {
                if let Some(expected_label) = self.func_def.param_label(db, i);
                if !expected_label.is_self(db.as_hir_db());
                if Some(expected_label) != given.label;
                then {
                    let diag = BodyDiag::CallArgLabelMismatch {
                        primary: given.label_span.unwrap_or(given.expr_span.clone()),
                        def_span: self.func_def.name_span(db),
                        given: given.label,
                        expected: expected_label,
                    };
                    tc.push_diag(diag);
                }
            }

            let expected = expected.instantiate(db, &self.generic_args);
            tc.equate_ty(given.expr_prop.ty, expected, given.expr_span);
        }
    }
}

/// The lowered representation of [`HirCallArg`]
struct CallArg<'db> {
    label: Option<IdentId<'db>>,
    expr_prop: ExprProp<'db>,
    label_span: Option<DynLazySpan<'db>>,
    expr_span: DynLazySpan<'db>,
}

impl<'db> CallArg<'db> {
    fn from_hir_arg(
        tc: &mut TyChecker<'db>,
        arg: &HirCallArg<'db>,
        span: LazyCallArgSpan<'db>,
    ) -> Self {
        let ty = tc.fresh_ty();
        let expr_prop = tc.check_expr(arg.expr, ty);
        let label = arg.label_eagerly(tc.db.as_hir_db(), tc.body());
        let label_span = arg.label.is_some().then(|| span.label().into());
        let expr_span = span.expr().into();

        Self::new(label, expr_prop, label_span, expr_span)
    }

    fn new(
        label: Option<IdentId<'db>>,
        expr_prop: ExprProp<'db>,
        label_span: Option<DynLazySpan<'db>>,
        expr_span: DynLazySpan<'db>,
    ) -> Self {
        Self {
            label,
            expr_prop,
            label_span,
            expr_span,
        }
    }
}
