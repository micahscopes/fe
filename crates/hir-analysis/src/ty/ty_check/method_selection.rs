use common::indexmap::IndexSet;
use either::Either;
use hir::{
    hir_def::{scope_graph::ScopeId, IdentId, Trait},
    span::DynLazySpan,
};
use itertools::Itertools;
use rustc_hash::FxHashSet;

use crate::{
    name_resolution::{available_traits_in_scope, diagnostics::NameResDiag, is_scope_visible_from},
    ty::{
        canonical::{Canonical, Canonicalized, Solution},
        diagnostics::{BodyDiag, FuncBodyDiag},
        fold::TyFoldable,
        func_def::FuncDef,
        method_table::probe_method,
        trait_def::{impls_for_ty, TraitDef, TraitInstId, TraitMethod},
        trait_lower::lower_trait,
        trait_resolution::{is_goal_satisfiable, GoalSatisfiability, PredicateListId},
        ty_def::TyId,
        unify::UnificationTable,
    },
    HirAnalysisDb,
};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(super) enum Candidate<'db> {
    InherentMethod(FuncDef<'db>),
    TraitMethod(TraitMethodCand<'db>),
    NeedsConfirmation(TraitMethodCand<'db>),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub(super) struct TraitMethodCand<'db> {
    pub(super) inst: Solution<TraitInstId<'db>>,
    pub(super) method: TraitMethod<'db>,
}

impl<'db> TraitMethodCand<'db> {
    fn new(inst: Solution<TraitInstId<'db>>, method: TraitMethod<'db>) -> Self {
        Self { inst, method }
    }
}

pub(super) fn select_method_candidate<'db>(
    db: &'db dyn HirAnalysisDb,
    receiver: (Canonical<TyId<'db>>, DynLazySpan<'db>),
    method_name: (IdentId<'db>, DynLazySpan<'db>),
    scope: ScopeId<'db>,
    assumptions: PredicateListId<'db>,
) -> Result<Candidate<'db>, FuncBodyDiag<'db>> {
    if receiver.0.value.is_ty_var(db) {
        return Err(BodyDiag::TypeMustBeKnown(method_name.1).into());
    }

    let candidates = assemble_method_candidates(db, receiver.0, method_name.0, scope, assumptions);

    let selector = MethodSelector {
        db,
        receiver: receiver.0,
        scope,
        candidates,
        assumptions,
    };

    match selector.select() {
        Ok(candidate) => Ok(candidate),

        Err(MethodSelectionError::AmbiguousInherentMethod(cands)) => {
            let cand_spans = cands.into_iter().map(|cand| cand.name_span(db)).collect();
            let diag = BodyDiag::AmbiguousInherentMethodCall {
                primary: method_name.1,
                method_name: method_name.0,
                cand_spans,
            };

            Err(diag.into())
        }

        Err(MethodSelectionError::AmbiguousTraitMethod(traits)) => {
            let traits = traits.into_iter().map(|def| def.trait_(db)).collect();

            let diag = BodyDiag::AmbiguousTrait {
                primary: method_name.1,
                method_name: method_name.0,
                traits,
            };

            Err(diag.into())
        }

        Err(MethodSelectionError::NotFound) => {
            let base_ty = receiver.0.value.base_ty(db);
            let diag =
                BodyDiag::method_not_found(db, method_name.1, method_name.0, Either::Left(base_ty));
            Err(diag.into())
        }

        Err(MethodSelectionError::InvisibleInherentMethod(func)) => {
            let diag =
                NameResDiag::Invisible(method_name.1, method_name.0, func.name_span(db).into());
            Err(diag.into())
        }

        Err(MethodSelectionError::InvisibleTraitMethod(traits)) => {
            let diag = BodyDiag::InvisibleAmbiguousTrait {
                primary: method_name.1,
                traits,
            };
            Err(diag.into())
        }
    }
}

fn assemble_method_candidates<'db>(
    db: &'db dyn HirAnalysisDb,
    receiver_ty: Canonical<TyId<'db>>,
    method_name: IdentId<'db>,
    scope: ScopeId<'db>,
    assumptions: PredicateListId<'db>,
) -> AssembledCandidates<'db> {
    CandidateAssembler {
        db,
        receiver_ty,
        method_name,
        scope,
        assumptions,
        candidates: AssembledCandidates::default(),
    }
    .assemble()
}

struct CandidateAssembler<'db> {
    db: &'db dyn HirAnalysisDb,
    /// The type that method is being called on.
    receiver_ty: Canonical<TyId<'db>>,
    /// The name of the method being called.
    method_name: IdentId<'db>,
    /// The scope that candidates are being assembled in.
    scope: ScopeId<'db>,
    /// The assumptions for the type bound in the current scope.
    assumptions: PredicateListId<'db>,
    candidates: AssembledCandidates<'db>,
}

impl<'db> CandidateAssembler<'db> {
    fn assemble(mut self) -> AssembledCandidates<'db> {
        self.assemble_inherent_method_candidates();
        self.assemble_trait_method_candidates();
        self.candidates
    }

    fn assemble_inherent_method_candidates(&mut self) {
        let ingot = self.scope.ingot(self.db.as_hir_db());
        for &method in probe_method(self.db, ingot, self.receiver_ty, self.method_name) {
            self.candidates.insert_inherent_method(method);
        }
    }

    fn assemble_trait_method_candidates(&mut self) {
        let ingot = self.scope.ingot(self.db.as_hir_db());
        let mut table = UnificationTable::new(self.db);
        let extracted_receiver_ty = self.receiver_ty.extract_identity(&mut table);

        for &implementor in impls_for_ty(self.db, ingot, self.receiver_ty) {
            let trait_def = implementor.skip_binder().trait_def(self.db);
            self.insert_trait_method_cand(trait_def)
        }

        for &pred in self.assumptions.list(self.db) {
            let snapshot = table.snapshot();
            let self_ty = pred.self_ty(self.db);
            let self_ty = table.instantiate_to_term(self_ty);

            if table.unify(extracted_receiver_ty, self_ty).is_ok() {
                self.insert_trait_method_cand(pred.def(self.db));
                for super_trait in pred.def(self.db).super_traits(self.db) {
                    let super_trait = super_trait.instantiate(self.db, pred.args(self.db));
                    self.insert_trait_method_cand(super_trait.def(self.db));
                }
            }

            table.rollback_to(snapshot);
        }
    }

    fn insert_trait_method_cand(&mut self, trait_def: TraitDef<'db>) {
        if let Some(&trait_method) = trait_def.methods(self.db).get(&self.method_name) {
            self.candidates.insert_trait(trait_def, trait_method);
        }
    }
}

struct MethodSelector<'db> {
    db: &'db dyn HirAnalysisDb,
    receiver: Canonical<TyId<'db>>,
    scope: ScopeId<'db>,
    candidates: AssembledCandidates<'db>,
    assumptions: PredicateListId<'db>,
}

impl<'db> MethodSelector<'db> {
    fn select(self) -> Result<Candidate<'db>, MethodSelectionError<'db>> {
        if let Some(res) = self.select_inherent_method() {
            return res;
        }

        self.select_trait_methods()
    }

    fn select_inherent_method(&self) -> Option<Result<Candidate<'db>, MethodSelectionError<'db>>> {
        let inherent_methods = &self.candidates.inherent_methods;
        let visible_inherent_methods: Vec<_> = inherent_methods
            .iter()
            .copied()
            .filter(|cand| self.is_inherent_method_visible(*cand))
            .collect();

        match visible_inherent_methods.len() {
            0 => {
                if inherent_methods.is_empty() {
                    None
                } else {
                    Some(Err(MethodSelectionError::InvisibleInherentMethod(
                        *inherent_methods.iter().next().unwrap(),
                    )))
                }
            }
            1 => Some(Ok(Candidate::InherentMethod(visible_inherent_methods[0]))),

            _ => Some(Err(MethodSelectionError::AmbiguousInherentMethod(
                inherent_methods.iter().copied().collect(),
            ))),
        }
    }

    /// Selects the most appropriate trait method candidate.
    ///
    /// This function checks the available trait method candidates and attempts
    /// to find the best match. If there is only one candidate, it is returned.
    /// If there are multiple candidates, it checks for visibility and
    /// ambiguity.
    ///
    /// **NOTE**: If there is no ambiguity, the trait does not need to be
    /// visible.
    ///
    /// # Returns
    ///
    /// * `Ok(Candidate)` - The selected method candidate.
    /// * `Err(MethodSelectionError)` - An error indicating the reason for
    ///   failure.
    fn select_trait_methods(&self) -> Result<Candidate<'db>, MethodSelectionError<'db>> {
        let traits = &self.candidates.traits;

        if traits.len() == 1 {
            let (def, method) = traits.iter().next().unwrap();
            return Ok(self.find_inst(*def, *method));
        }

        let available_traits = self.available_traits();
        let visible_traits: Vec<_> = traits
            .iter()
            .copied()
            .filter(|cand| available_traits.contains(&cand.0))
            .collect();

        match visible_traits.len() {
            0 => {
                if traits.is_empty() {
                    Err(MethodSelectionError::NotFound)
                } else {
                    // Suggests trait imports.
                    let traits = traits.iter().map(|(def, _)| def.trait_(self.db)).collect();
                    Err(MethodSelectionError::InvisibleTraitMethod(traits))
                }
            }

            1 => {
                let (def, method) = visible_traits[0];
                Ok(self.find_inst(def, method))
            }

            _ => Err(MethodSelectionError::AmbiguousTraitMethod(
                visible_traits.into_iter().map(|cand| cand.0).collect(),
            )),
        }
    }

    /// Finds an instance of a trait method for the given trait definition and
    /// method.
    ///
    /// This function attempts to unify the receiver type with the method's self
    /// type, and assigns type variables to the trait parameters. It then
    /// checks if the goal is satisfiable given the current assumptions.
    /// Depending on the result, it either returns a confirmed trait method
    /// candidate or one that needs further confirmation.
    ///
    /// # Arguments
    ///
    /// * `def` - The trait definition.
    /// * `method` - The trait method.
    ///
    /// # Returns
    ///
    /// A `Candidate` representing the found trait method instance.
    fn find_inst(&self, def: TraitDef<'db>, method: TraitMethod<'db>) -> Candidate<'db> {
        let mut table = UnificationTable::new(self.db);
        let receiver = self.receiver.extract_identity(&mut table);

        // Assign type variables to trait parameters.
        let inst_args = def
            .params(self.db)
            .iter()
            .map(|ty| table.new_var_from_param(*ty))
            .collect_vec();

        let cand = TraitInstId::new(self.db, def, inst_args);
        // Unify receiver and method self.
        method.instantiate_with_inst(&mut table, receiver, cand);

        let cand = cand.fold_with(&mut table);
        let canonical_cand = Canonicalized::new(self.db, cand);

        match is_goal_satisfiable(
            self.db,
            self.scope.ingot(self.db.as_hir_db()),
            canonical_cand.value,
            self.assumptions,
        ) {
            GoalSatisfiability::Satisfied(solution) => {
                // Map back the solution to the current context.
                let solution = canonical_cand.extract_solution(&mut table, *solution);

                // Unify candidate to solution.
                table.unify(cand, solution).unwrap();

                Candidate::TraitMethod(TraitMethodCand::new(
                    self.receiver
                        .canonicalize_solution(self.db, &mut table, cand),
                    method,
                ))
            }

            &GoalSatisfiability::NeedsConfirmation(_)
            | GoalSatisfiability::ContainsInvalid
            | GoalSatisfiability::UnSat(_) => Candidate::NeedsConfirmation(TraitMethodCand::new(
                self.receiver
                    .canonicalize_solution(self.db, &mut table, cand),
                method,
            )),
        }
    }

    fn is_inherent_method_visible(&self, def: FuncDef) -> bool {
        is_scope_visible_from(self.db, def.scope(self.db), self.scope)
    }

    fn available_traits(&self) -> FxHashSet<TraitDef<'db>> {
        let mut traits = FxHashSet::default();

        let mut insert_trait = |trait_def: TraitDef<'db>| {
            traits.insert(trait_def);

            for trait_ in trait_def.super_traits(self.db) {
                traits.insert(trait_.skip_binder().def(self.db));
            }
        };

        for &trait_ in available_traits_in_scope(self.db, self.scope) {
            let trait_def = lower_trait(self.db, trait_);
            insert_trait(trait_def);
        }

        for pred in self.assumptions.list(self.db) {
            let trait_def = pred.def(self.db);
            insert_trait(trait_def)
        }

        traits
    }
}

pub enum MethodSelectionError<'db> {
    AmbiguousInherentMethod(Vec<FuncDef<'db>>),
    AmbiguousTraitMethod(Vec<TraitDef<'db>>),
    NotFound,
    InvisibleInherentMethod(FuncDef<'db>),
    InvisibleTraitMethod(Vec<Trait<'db>>),
}

#[derive(Default)]
struct AssembledCandidates<'db> {
    inherent_methods: FxHashSet<FuncDef<'db>>,
    traits: IndexSet<(TraitDef<'db>, TraitMethod<'db>)>,
}

impl<'db> AssembledCandidates<'db> {
    fn insert_inherent_method(&mut self, method: FuncDef<'db>) {
        self.inherent_methods.insert(method);
    }

    fn insert_trait(&mut self, def: TraitDef<'db>, method: TraitMethod<'db>) {
        self.traits.insert((def, method));
    }
}
