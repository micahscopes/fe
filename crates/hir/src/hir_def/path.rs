use super::{GenericArgListId, IdentId};
use crate::{hir_def::Partial, HirDb};

#[salsa::interned]
pub struct PathId<'db> {
    pub ident: Partial<IdentId<'db>>,
    pub generic_args: GenericArgListId<'db>,
    pub parent: Option<PathId<'db>>,
}

impl<'db> PathId<'db> {
    pub fn from_ident(db: &'db dyn HirDb, ident: IdentId<'db>) -> Self {
        Self::new(
            db,
            Partial::Present(ident),
            GenericArgListId::none(db),
            None,
        )
    }

    pub fn self_ty(db: &'db dyn HirDb) -> Self {
        Self::from_ident(db, IdentId::make_self_ty(db))
    }

    pub fn len(self, db: &dyn HirDb) -> usize {
        if let Some(parent) = self.parent(db) {
            parent.len(db) + 1
        } else {
            1
        }
    }

    pub fn segment(self, db: &'db dyn HirDb, idx: usize) -> Option<PathId<'db>> {
        if idx == self.len(db) - 1 {
            Some(self)
        } else {
            self.parent(db).and_then(|p| p.segment(db, idx))
        }
    }

    pub fn root_ident(self, db: &'db dyn HirDb) -> Option<IdentId<'db>> {
        if let Some(parent) = self.parent(db) {
            parent.root_ident(db)
        } else {
            self.ident(db).to_opt()
        }
    }

    // xxx check uses, remove?
    pub fn as_ident(self, db: &'db dyn HirDb) -> Option<IdentId<'db>> {
        if self.parent(db).is_none() && self.generic_args(db).is_empty(db) {
            self.ident(db).to_opt()
        } else {
            None
        }
    }

    pub fn push(
        self,
        db: &'db dyn HirDb,
        ident: Partial<IdentId<'db>>,
        generic_args: GenericArgListId<'db>,
    ) -> Self {
        Self::new(db, ident, generic_args, Some(self))
    }

    pub fn push_ident(self, db: &'db dyn HirDb, ident: IdentId<'db>) -> Self {
        Self::new(
            db,
            Partial::Present(ident),
            GenericArgListId::none(db),
            Some(self),
        )
    }

    pub fn pretty_print(self, db: &dyn HirDb) -> String {
        let ident = self.ident(db).to_opt().map_or("_", |id| id.data(db));
        if let Some(parent) = self.parent(db) {
            parent.pretty_print(db) + "::" + ident
        } else {
            ident.to_string()
        }
    }
}
