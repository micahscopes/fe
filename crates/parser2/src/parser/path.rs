use std::convert::identity;

use crate::{ParseError, SyntaxKind};

use super::{
    define_scope,
    expr::{is_lshift, is_lt_eq},
    param::GenericArgListScope,
    token_stream::TokenStream,
    Parser,
};

define_scope! {
    #[doc(hidden)]
    pub PathScope { is_expr: bool },
    Path,
    (Colon2)
}
impl super::Parse for PathScope {
    type Error = ParseError;

    fn parse<S: TokenStream>(&mut self, parser: &mut Parser<S>) -> Result<(), Self::Error> {
        parser.set_newline_as_trivia(false);
        parser.parse(PathSegmentScope::new(self.is_expr))?;
        while parser.bump_if(SyntaxKind::Colon2) {
            parser.parse(PathSegmentScope::default())?;
        }
        Ok(())
    }
}

define_scope! { PathSegmentScope { is_expr: bool }, PathSegment }
impl super::Parse for PathSegmentScope {
    type Error = ParseError;

    fn parse<S: TokenStream>(&mut self, parser: &mut Parser<S>) -> Result<(), Self::Error> {
        match parser.current_kind() {
            Some(kind) if is_path_segment(kind) => {
                parser.bump();

                if parser.current_kind() == Some(SyntaxKind::Lt)
                    && !(is_lt_eq(parser) || is_lshift(parser))
                    && parser.dry_run(|parser| {
                        parser
                            .parse_ok(GenericArgListScope::new(self.is_expr))
                            .is_ok_and(identity)
                    })
                {
                    parser
                        .parse(GenericArgListScope::new(self.is_expr))
                        .expect("dry_run suggests this will succeed");
                }
                Ok(())
            }
            _ => Err(ParseError::expected(
                &[SyntaxKind::PathSegment],
                None,
                parser.end_of_prev_token,
            )),
        }
    }
}

pub(super) fn is_path_segment(kind: SyntaxKind) -> bool {
    matches!(
        kind,
        SyntaxKind::SelfTypeKw
            | SyntaxKind::SelfKw
            | SyntaxKind::IngotKw
            | SyntaxKind::SuperKw
            | SyntaxKind::Ident
    )
}
