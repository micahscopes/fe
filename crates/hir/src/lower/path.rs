use crate::hir_def::{GenericArgListId, IdentId, Partial, PathId};
use parser::ast::{self, GenericArgsOwner};

use super::FileLowerCtxt;

impl<'db> PathId<'db> {
    pub(super) fn lower_ast(ctxt: &mut FileLowerCtxt<'db>, ast: ast::Path) -> Self {
        let db = ctxt.db();

        let mut path: Option<Self> = None;
        for seg in ast.into_iter() {
            let ident = match seg.kind() {
                Some(ast::PathSegmentKind::Ingot(_)) => Some(IdentId::make_ingot(db)),
                Some(ast::PathSegmentKind::Super(_)) => Some(IdentId::make_super(db)),
                Some(ast::PathSegmentKind::SelfTy(_)) => Some(IdentId::make_self_ty(db)),
                Some(ast::PathSegmentKind::Self_(_)) => Some(IdentId::make_self(db)),
                Some(ast::PathSegmentKind::Ident(ident)) => Some(IdentId::lower_token(ctxt, ident)),
                None => None,
            }
            .into();

            let generic_args = GenericArgListId::lower_ast_opt(ctxt, seg.generic_args());

            path = path
                .map(|p| p.push(db, ident, generic_args))
                .or_else(|| Some(Self::new(db, ident, generic_args, None)))
        }

        path.expect("ast::Path must contain at least 1 segment")
    }

    pub(super) fn lower_ast_partial(
        ctxt: &mut FileLowerCtxt<'db>,
        ast: Option<ast::Path>,
    ) -> Partial<Self> {
        ast.map(|ast| Self::lower_ast(ctxt, ast)).into()
    }
}
