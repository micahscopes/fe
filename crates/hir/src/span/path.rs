use parser::ast;

use super::{define_lazy_span_node, params::LazyGenericArgListSpan, LazySpanAtom};

define_lazy_span_node!(
    LazyPathSpan,
    ast::Path,
    @idx {
        (segment, LazyPathSegmentSpan),
    }
);

define_lazy_span_node!(
    LazyPathSegmentSpan,
    ast::PathSegment,
    @token {
        (ident, ident),
    }
    @node {
        (generic_args, generic_args, LazyGenericArgListSpan),
    }
);
impl<'db> LazyPathSegmentSpan<'db> {
    pub fn into_atom(self) -> LazySpanAtom<'db> {
        LazySpanAtom(self.0)
    }
}
