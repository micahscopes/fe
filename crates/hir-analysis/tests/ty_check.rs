mod test_db;
use std::path::Path;

use dir_test::{dir_test, Fixture};
use fe_compiler_test_utils::snap_test;
use fe_hir_analysis::ty::ty_check::check_func_body;
use test_db::HirAnalysisTestDb;

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_files/ty_check",
    glob: "**/*.fe"
)]
fn ty_check_standalone(fixture: Fixture<&str>) {
    let mut db = HirAnalysisTestDb::default();
    let path = Path::new(fixture.path());
    let file_name = path.file_name().and_then(|file| file.to_str()).unwrap();
    let input = db.new_stand_alone(file_name, fixture.content());
    let (top_mod, mut prop_formatter) = db.top_mod(input);

    db.assert_no_diags(top_mod);

    for &func in top_mod.all_funcs(&db) {
        let Some(body) = func.body(&db) else {
            continue;
        };

        let typed_body = &check_func_body(&db, func).1;
        for expr in body.exprs(&db).keys() {
            let ty = typed_body.expr_ty(&db, expr);
            prop_formatter.push_prop(
                func.top_mod(&db),
                expr.lazy_span(body).into(),
                ty.pretty_print(&db).to_string(),
            );
        }

        for pat in body.pats(&db).keys() {
            let ty = typed_body.pat_ty(&db, pat);
            prop_formatter.push_prop(
                func.top_mod(&db),
                pat.lazy_span(body).into(),
                ty.pretty_print(&db).to_string(),
            );
        }
    }

    let res = prop_formatter.finish(&db);
    snap_test!(res, fixture.path());
}
