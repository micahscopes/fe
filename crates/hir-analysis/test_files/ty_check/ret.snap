---
source: crates/hir-analysis/tests/ty_check.rs
expression: res
input_file: crates/hir-analysis/test_files/ty_check/ret.fe
---
note: 
   ┌─ ret.fe:7:35
   │  
 7 │   fn foo(b1: bool, b2: bool) -> i32 {
   │ ╭───────────────────────────────────^
 8 │ │     use Tag::*
 9 │ │ 
10 │ │     let tag = if b1 {
   · │
36 │ │     }
37 │ │ }
   │ ╰─^ i32

note: 
   ┌─ ret.fe:10:9
   │
10 │     let tag = if b1 {
   │         ^^^ Tag

note: 
   ┌─ ret.fe:10:15
   │  
10 │       let tag = if b1 {
   │ ╭───────────────^
11 │ │         return 0
12 │ │     } else if b2 {
13 │ │         Tag1
14 │ │     } else {
15 │ │         Tag3
16 │ │     }
   │ ╰─────^ Tag

note: 
   ┌─ ret.fe:10:18
   │
10 │     let tag = if b1 {
   │                  ^^ bool

note: 
   ┌─ ret.fe:10:21
   │  
10 │       let tag = if b1 {
   │ ╭─────────────────────^
11 │ │         return 0
12 │ │     } else if b2 {
   │ ╰─────^ !

note: 
   ┌─ ret.fe:11:16
   │
11 │         return 0
   │                ^ i32

note: 
   ┌─ ret.fe:12:12
   │  
12 │       } else if b2 {
   │ ╭────────────^
13 │ │         Tag1
14 │ │     } else {
15 │ │         Tag3
16 │ │     }
   │ ╰─────^ Tag

note: 
   ┌─ ret.fe:12:15
   │
12 │     } else if b2 {
   │               ^^ bool

note: 
   ┌─ ret.fe:12:18
   │  
12 │       } else if b2 {
   │ ╭──────────────────^
13 │ │         Tag1
14 │ │     } else {
   │ ╰─────^ Tag

note: 
   ┌─ ret.fe:13:9
   │
13 │         Tag1
   │         ^^^^ Tag

note: 
   ┌─ ret.fe:14:12
   │  
14 │       } else {
   │ ╭────────────^
15 │ │         Tag3
16 │ │     }
   │ ╰─────^ Tag

note: 
   ┌─ ret.fe:15:9
   │
15 │         Tag3
   │         ^^^^ Tag

note: 
   ┌─ ret.fe:18:9
   │
18 │     let b = match tag {
   │         ^ bool

note: 
   ┌─ ret.fe:18:13
   │  
18 │       let b = match tag {
   │ ╭─────────────^
19 │ │         Tag1 => {
20 │ │             true
21 │ │         }
   · │
29 │ │         }
30 │ │     }
   │ ╰─────^ bool

note: 
   ┌─ ret.fe:18:19
   │
18 │     let b = match tag {
   │                   ^^^ Tag

note: 
   ┌─ ret.fe:19:9
   │
19 │         Tag1 => {
   │         ^^^^ Tag

note: 
   ┌─ ret.fe:19:17
   │  
19 │           Tag1 => {
   │ ╭─────────────────^
20 │ │             true
21 │ │         }
   │ ╰─────────^ bool

note: 
   ┌─ ret.fe:20:13
   │
20 │             true
   │             ^^^^ bool

note: 
   ┌─ ret.fe:23:9
   │
23 │         Tag2 => {
   │         ^^^^ Tag

note: 
   ┌─ ret.fe:23:17
   │  
23 │           Tag2 => {
   │ ╭─────────────────^
24 │ │             return 3
25 │ │         }
   │ ╰─────────^ bool

note: 
   ┌─ ret.fe:24:20
   │
24 │             return 3
   │                    ^ i32

note: 
   ┌─ ret.fe:27:9
   │
27 │         Tag3 => {
   │         ^^^^ Tag

note: 
   ┌─ ret.fe:27:17
   │  
27 │           Tag3 => {
   │ ╭─────────────────^
28 │ │             false
29 │ │         }
   │ ╰─────────^ bool

note: 
   ┌─ ret.fe:28:13
   │
28 │             false
   │             ^^^^^ bool

note: 
   ┌─ ret.fe:32:5
   │  
32 │ ╭     if b {
33 │ │         1
34 │ │     } else {
35 │ │         2
36 │ │     }
   │ ╰─────^ i32

note: 
   ┌─ ret.fe:32:8
   │
32 │     if b {
   │        ^ bool

note: 
   ┌─ ret.fe:32:10
   │  
32 │       if b {
   │ ╭──────────^
33 │ │         1
34 │ │     } else {
   │ ╰─────^ i32

note: 
   ┌─ ret.fe:33:9
   │
33 │         1
   │         ^ i32

note: 
   ┌─ ret.fe:34:12
   │  
34 │       } else {
   │ ╭────────────^
35 │ │         2
36 │ │     }
   │ ╰─────^ i32

note: 
   ┌─ ret.fe:35:9
   │
35 │         2
   │         ^ i32
